Improve project organization

shykes opened the pull request:
Proposed improvements to project organization.

This organization change has 3 goals:

1) Make the project more open and accessible
2) Make the project more scalable
3) Do this in an incremental way, without unnecessary refactoring

The actual organization change is covered in 6a99b8dea935be9228a47c84bcbfb7491e6e5584. Cosmetic changes are discussed in #9177 and #9178

Commit:
Proposal for an improved project structure.

Note: this deprecates the fine-grained, high-overlap cascading MAINTAINERS files,
and replaces them with a single top-level file, using a new structure:

* More coarse grained subsystems with dedicated teams of maintainers
* Core maintainers with a better-defined role and a wider scope (if it's
not in a subsystem, it's up to the core maintainers to figure it out)
* Architects
* Operators

This is work in progress, the goal is to start a conversation

Signed-off-by: Solomon Hykes <solomon@docker.com>
Signed-off-by: Erik Hollensbe <github@hollensbe.org>
Signed-off-by: Arnaud Porterie <arnaud.porterie@docker.com>
Signed-off-by: Tibor Vass <teabee89@gmail.com>
Signed-off-by: Victor Vieux <vieux@docker.com>
Signed-off-by: Vincent Batts <vbatts@redhat.com>

vbatts commented:
@shykes next steps? and DCO marker? ;-)

tiborvass commented on MAINTAINERS:
```diff
@@ -1,9 +1,456 @@
+			"lkd4d"
+			"tibor",
+			"vbatts"
+			"vieux",
```
vbatts misses comma, and vieux should not have one

icecrime commented on MAINTAINERS:
```diff
@@ -1,9 +1,462 @@
+			"erikh",
+			"icecrime",
+			"jfrazelle",
+			"lkd4d"
```
It's "lk4d4"!

icecrime commented on MAINTAINERS:
```diff
@@ -1,9 +1,462 @@
+
+	# As the project grows, it gets separated into well-defined subsystems. Each subsystem
+	# has a dedicated group of maintainers, which are responsible for its quali
+	# dedicated to that subsytem. This "cellular division" is the primary mechanism
```
Something wrong here (truncated word "quali" unrelated to the end of the sentence).

tiborvass commented:
Overall LGTM :)

icecrime commented:
Same here, I don't see any remaining issues: LGTM :-1:

unclejack commented:
LGTM

jfrazelle commented:
LGTM

shykes commented:
Ok, after much discussion on irc, I think we're ready to merge.

* Added missing people
* Added missing subsystems
* Officialized de-facto core maintainers
* Removed the vague notion of "lead"
* Clarified the roles of Chief Architect, Chief Maintainer, Chief Operator

Even merged, this doesn't end the conversation. The goal is to formalize a structure which is already mostly in place informally. This is a necessary starting point for attacking more ambitious upgrades to the organization.

The big topic which remains to be addressed: how do we transform from a tool-centric organization to a platform-centric organization? Docker is no longer a single tool, but a collection of tools unified under a platform. We want this platform to be composable (each tool should be useful standalone, and replaceable by another), but also well integrated: the user experience should feel like more than the sum of the parts. To achieve this, we need to address *integration* at many levels. More on this topic soon!

shykes commented:
@cyphar it's mostly formalizing what is already in place, but also incorporating de-facto rules that have appeared organically but weren't documented. For example, the concept of "core maintainers" which are the de facto owners of everything which isn't a clearly separated subsystem - that was already the case but not clearly documented. This file also deprecates the more fine-grained MAINTAINERS files - because in practice we were ignoring them often.

The other important change is that I am reducing the role of the bdfl, and creating 3 new leadership roles: Chief Architect, Chief Maintainer, Chief Operator. That should help reduce the dependency on bdfl for running the project, while keeping it as a guarantee for the long-term direction and governance of the project.
